Admin panel list displays improvement for Property (harvest app) and Person (member app) - new

farhanmasud opened the pull request:
Added relevant fields in list_display, list_filter and search_fields for Property (harvest app) and Person (member app) in Django admin.

Links to check after running locally -

[http://localhost:8000/admin/harvest/property/](http://localhost:8000/admin/harvest/property/)
[http://localhost:8000/admin/member/person/](http://localhost:8000/admin/member/person/)

Screenshots -

For Property model -

![property-new](https://user-images.githubusercontent.com/43474661/160253225-181e5832-b2f2-423a-9112-8a9a25b63f4b.png)

For Person model -

![person-new](https://user-images.githubusercontent.com/43474661/160253240-9c69e77b-49e6-418e-a9bc-e50205a1a7ff.png)

Intended fix for [#119](https://github.com/LesFruitsDefendus/saskatoon-ng/issues/119)

Also made the change to keep Django version under 4.0.0 to solve ImportError of 'force_text' from 'django.utils.encoding'.

tristanlatr commented:
Works great!
Thanks!

2ynn commented:
@tristanlatr please let's not merge this as is, see comments above

tristanlatr commented:
Which comment @2ynn ?

2ynn commented:
@tristanlatr sorry didn't realize comments on code were in a pending review

farhanmasud commented:
@2ynn thank you very much for your valuable feedback. I'll make the changes accordingly and will update as soon as possible

farhanmasud commented:
@2ynn I have removed the fields that are not relevant at this moment. Currently there are both `postal_code` and `neighborhood` fields are in `list_filter` which might get long as the People and Property model instances grow.

Currently both of `postal_code` and `neighborhood` are in the `list_filter` for both models which might grow long as the model instances are being added. Should I remove `postal_code` from the `list_filter` and add it to the `search_fields`?

We can also implement a custom search [(Ref.)](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_search_results) for postal_code by cleaning up the input search lookup data (for example removing the space from a postal code and matching that up against the database `postal_code` by doing the same there).

2ynn commented on saskatoon/harvest/admin.py:
```diff
@@ -66,6 +66,39 @@ class PropertyImageInline(admin.TabularInline):
+    search_fields = (
+        "street_number",
+        "street",
+        "complement",
+        "publishable_location",
+        "additional_info",
```
I don't think these are relevant. I would rather be able to search info about the owner, e.g. `owner__person__last_name` or `owner__person__authuser__email`

2ynn commented on saskatoon/member/admin.py:
```diff
@@ -205,11 +205,39 @@ def deactivate_account(self, request, queryset):
+        "first_name",
+        "family_name",
+        "phone",
+        "street_number",
+        "street",
+        "complement",
+        "comments",
```
same here, i'd rather be able to search `authuser__email` for example

2ynn commented:
@farhanmasud yes searching for postal code would be best, please remove it from filters. Postal codes are currently simple CharFields in the database and were not validated so formatting can be inconsistent throughout.

farhanmasud commented:
@2ynn I'm updating all accordingly

farhanmasud commented:
@2ynn @tristanlatr I have updated the `list_filter` and `search_fields` as discussed.

I have also implemented a custom search lookup for `postal_code` field by using [get_queryset](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_queryset) method instead of the [get_search_results](https://docs.djangoproject.com/en/3.2/ref/contrib/admin/#django.contrib.admin.ModelAdmin.get_search_results) method that I initially referenced. Here I have only cleaned the field by removing spaces as Django admin search automatically converts all search fields to lowercase.

Please let me know if there are any other modifications needed with the `list_filters` and this custom `postal_code` field lookup that I have implemented.

2ynn commented on saskatoon/harvest/admin.py:
```diff
@@ -66,8 +68,52 @@ class PropertyImageInline(admin.TabularInline):
     model = Property
     inlines = [PropertyImageInline]
+    list_display = (
+        '__str__',
```
`__str__` does not include postal_code at the moment, please add the postal_code to list_display

2ynn commented on saskatoon/harvest/admin.py:
```diff
@@ -66,8 +68,52 @@ class PropertyImageInline(admin.TabularInline):
+        'authorized',
+        'pending',
+        'harvest_every_year', 
+        'trees_location',
+        'avg_nb_required_pickers',
```
these two become way too long, let's remove them

2ynn commented on saskatoon/harvest/admin.py:
```diff
@@ -66,8 +68,52 @@ class PropertyImageInline(admin.TabularInline):
     exclude = ['longitude', 'latitude', 'geom']
 
+    def get_queryset(self, request):
+        queryset = super().get_queryset(request)
+        queryset = queryset.annotate(
+            postal_code_cleaned=Replace(
+                'postal_code',
+                Value(" "),
+                Value("")
+            )
```
nice!

2ynn commented on saskatoon/member/admin.py:
```diff
@@ -205,11 +207,49 @@ def deactivate_account(self, request, queryset):
+                Value("")
+            )
+        )
+        
+        return queryset
```
nitty gritty: can we try avoiding committing extra whitespaces in the file? 
For example I see one on this empty line with my editor.
Really no big deal though!
Btw if you want we could go shorter with :
```
def get_queryset(self, request):
        queryset = super().get_queryset(request)
        return queryset.annotate(
            postal_code_cleaned=Replace(
                'postal_code',
                Value(" "),
                Value("")
            )
        )
```

2ynn reviewed (commented):
Thanks for the changes, postal_code search works great!

farhanmasud commented:
@2ynn I'm updating all accordingly. Thanks for the feedback!

farhanmasud commented:
@2ynn I have added `postal_code` field on the `list_display` attribute of both of `PropertyAdmin` and `PersonAdmin`.
Removed fields `trees_location` and `avg_nb_required_pickers` fields from `PropertyAdmin`.
And reformatted the `get_queryset` method by removing the extra whitespace and making it shorter.

tristanlatr commented:
Thanks for the great work!
